### PR TITLE
Fix S2259 FP: IsType returns false for Null values

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Follow up of #6986

### CODE REVIEW

File: analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json [hunk 1]
```diff
@@ -2,19 +2,6 @@
 "issues":  [
 {
 "id":  "S2259",
-"message":  "'deadLetter' is null on at least one execution path.",
-"location":  {
-"uri":  "sources\akka.net\src\core\Akka.Remote\RemoteActorRefProvider.cs",
-"region":  {
-"startLine":  802,
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/akka.net/Akka.Remote--netstandard2.0-S2259.json:9]: Looks good:
https://github.com/SonarSource/sonar-dotnet/blob/9552b1d6f148cd5c6472323f02eaf6d210d7dd42/analyzers/its/sources/akka.net/src/core/Akka.Remote/RemoteActorRefProvider.cs#L800-L802

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json [hunk 2]
```diff
@@ -1,17 +0,0 @@
-{
-"issues":  [
-{
-"id":  "S2259",
-"message":  "'module' is null on at least one execution path.",
-"location":  {
-"uri":  "sources\akka.net\src\core\Akka.Streams\Implementation\Fusing\StreamOfStreams.cs",
-"region":  {
-"startLine":  1217,
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S2259.json:9]: Also good:
https://github.com/SonarSource/sonar-dotnet/blob/9552b1d6f148cd5c6472323f02eaf6d210d7dd42/analyzers/its/sources/akka.net/src/core/Akka.Streams/Implementation/Fusing/StreamOfStreams.cs#L1215-L1217

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsType.cs [hunk 3]
```diff
@@ -28,7 +28,7 @@ internal sealed class IsType : BranchingProcessor<IIsTypeOperationWrapper>
         IIsTypeOperationWrapper.FromOperation(operation);
 
     protected override SymbolicConstraint BoolConstraintFromOperation(ProgramState state, IIsTypeOperationWrapper operation) =>
-        null;
+        state[operation.ValueOperand]?.HasConstraint(ObjectConstraint.Null) is true ? BoolConstraint.False : null;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsType.cs:31]: We could also learn true if ValueOperand has the "NotNullConstraint" and the type of the value is implicitly convertible to the type tested against. I think we have something similar in pattern handling.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 4]
```diff
@@ -393,6 +393,15 @@ public void Branching_LearnsObjectConstraint_IsType_Negated_CS(string expression
                 .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
         }
 
+        [TestMethod]
+        public void Branching_LearnsObjectConstraint_IsType_NullValue()
+        {
+            var validator = CreateIfElseEndValidatorCS("(object)null is Exception", OperationKind.IsType);  // Check something that is known to be null
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:399]: Add a test cases with a nullable value type being null:
```cs
(int?)null is object
(int?)null is int
(int?)null is int?
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good. We could also learn in the NotNull case (could be a follow-up PR). Test cases for nullable value types should be added.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsType.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsType.cs:31]: Yes, we could. I wouldn't go out of scope here. I don't see too many use cases for checking something that could be implicitly convertible.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**